lib/charms/tempo_k8s/v0/snapshot.py

```diff
@@ -0,0 +1,187 @@
+"""Snapshot charm library.
+
+This library acts as a bridge to scenario's
+[State](https://github.com/canonical/ops-scenario/blob/main/scenario/state.py#L861) data structure,
+allowing you to capture snapshots of a live charm's current state and serialize them as json.
+"""
+import logging
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, List, cast
+
+import ops
+import scenario
+import scenario.runtime
+
+if TYPE_CHECKING:
+    from scenario.state import AnyRelation
+
+logger = logging.getLogger("snapshot")
+
+# The unique Charmhub library identifier, never change it
+LIBID = "012181337a3a4717861b5fb4d8bdfc2b"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+PYDEPS = ["ops-scenario>=6"]
+
+
+class SnapshotBackend:
+    """Snapshot backend."""
+
+    def __init__(self, charm: ops.CharmBase):
+        self.charm = charm
+        self._unit_state_db = scenario.runtime.UnitStateDB(charm.charm_dir / ".unit-state.db")
+
+    def get_config(self) -> Dict:  # noqa: D102
+        return dict(self.charm.config)
+
+    def get_opened_ports(self) -> List[scenario.Port]:  # noqa: D102
+        return [scenario.Port(p.protocol, p.port) for p in self.charm.unit.opened_ports()]
+
+    def get_leader(self) -> bool:  # noqa: D102
+        return self.charm.unit.is_leader()
+
+    def get_model(self) -> scenario.Model:  # noqa: D102
+        return scenario.Model(
+            name=self.charm.model.name,
+            uuid=self.charm.model.uuid,
+            # fixme: how do we determine if we're on a kubernetes or machine model?
+        )
+
+    def get_planned_units(self) -> int:  # noqa: D102
+        return self.charm.app.planned_units()
+
+    def get_app_status(self) -> ops.StatusBase:  # noqa: D102
+        if not self.charm.unit.is_leader():
+            return ops.UnknownStatus()
+        return self.charm.app.status
+
+    def get_unit_status(self) -> ops.StatusBase:  # noqa: D102
+        return self.charm.unit.status
+
+    def get_workload_version(self) -> str:  # noqa: D102
+        return "42.todo"
```

Comment:
That one likely can disappear already

Reply:
yeah, turns out a charm can't access its own workload version